### PR TITLE
Validate peer_update in rust code

### PR DESCRIPTION
**Pull request by isoosiss7**
# Description
Define Channel, add pushUpdate and updateChannel

https://github.com/statechannels/native-utils/issues/18
https://github.com/statechannels/native-utils/issues/19
https://www.notion.so/statechannels/Rust-SC-library-93e36a45aeb8495f8c0325e6423eb86b#e6fff96fcb344469a258121bd19deffa

Work in progress...

### CODE REVIEW

File: packages/native-utils/common/src/channel.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    pub turn_num: Uint48,
+    pub is_final: bool,
+    pub outcome: Outcome,
+    pub app_data: Bytes,
+    pub hash: Bytes32,
+    pub signature: RecoverableSignature
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
```
Comment (andrewgordstewart) [inline packages/native-utils/common/src/channel.rs:22]: Rust is pretty fancy!

**Review by andrewgordstewart — COMMENTED** [review]
I left some comments, but this looks great so far!

I don't think you're looking for an approval, as you mentioned that this is a draft. For instance, the logic is currently missing the transition validation by the app definition -- I guess that's coming up soon?

We use [`pure-evm` for in javascript.](https://github.com/statechannels/statechannels/blob/85f923c7d51b2ca362537642d6455b35cdfa30c8/packages/server-wallet/src/evm-validator.ts#L29-L32). ([`pure-evm`](https://github.com/armaniferrante/pure-evm/blob/master/src/lib.rs) is a wrapper around a lot of the rust utilities we're using in this repository already. The fork that we depend on addresses an issue that `pure-evm` doesn't handle WRT pure functions -- I think signature checking doesn't work properly? I'll have to dig that up.)

File: packages/native-utils/common/src/state.rs [hunk 2]
```diff
@@ -219,12 +197,110 @@ impl State {
...
+    let parsed_signature = Signature::parse_slice(&signature[0..signature.len() - 1])
+        .or_else(|_| Err("invalid signature length"))?;
+    let recovery_id = RecoveryId::parse_rpc(signature[signature.len() - 1])
+        .or_else(|_| Err("invalid recovery ID"))?;
+    let public_key = recover(&message, &parsed_signature, &recovery_id)
+        .or_else(|_| Err("invalid signature"))?;
+
+    let recovered_address = checksum_address(public_key_to_address(public_key));
+
+    Ok(recovered_address.eq(&address))
```
[the remark this answers is not in the document]
Comment (isoosiss7) [reply packages/native-utils/common/src/state.rs]: Agree

I removed verify_sig, only do signature inside a state is enough so far.

File: packages/native-utils/common/src/state.rs [hunk 3]
```diff
@@ -219,12 +197,110 @@ impl State {
...
+                Err("Outcome change verboten")
+            } else {
+                Ok(Status::True)
+            }
+        } else {
+            if self.is_final {
+                Err("isFinal retrograde")
+            } else {
+                if to_state.turn_num < Uint48(2 * to_state.channel.participants.len() as u64) {
+                    if self.outcome != to_state.outcome {
```
[the remark this answers is not in the document]
Comment (isoosiss7) [reply packages/native-utils/common/src/state.rs:248]: :)
I added a test case to verify this works.

File: packages/native-utils/common/src/state.rs [hunk 4]
```diff
@@ -219,12 +197,110 @@ impl State {
...
+            Ok(())
+        }
+    }
+
+    pub fn require_valid_protocol_transition(&self, to_state: State) -> Result<Status, &'static str> {
+        self._require_extra_implicit_checks(&to_state)?;
+
+        if to_state.is_final {
+            if self.outcome != to_state.outcome {
+                Err("Outcome change verboten")
```
[the remark this answers is not in the document]
Comment (isoosiss7) [reply packages/native-utils/common/src/state.rs]: 'verboten' is borrowed from statechannel error message.
After consideration I changed it to 'forbidden' to make it easier for non-native english speakers like myself :)

File: packages/native-utils/tests/sign.test.ts [hunk 5]
```diff
@@ -88,19 +89,86 @@ describe('Sign state', () => {
...
-    const wasmSignature = wasm.signState(state, PRIVATE_KEY).signature
+    const wasmSigned = wasm.signState(state, PRIVATE_KEY1)
 
-    expect(nativeSignature).toStrictEqual(oldSignature)
-    expect(wasmSignature).toStrictEqual(oldSignature)
+    expect(nativeSigned.signature).toStrictEqual(oldSignature)
+    expect(wasmSigned.signature).toStrictEqual(oldSignature)
   })
 
+  test('Peer states validate as expected', async () => {
```
[the remark this answers is not in the document]
Comment (isoosiss7) [reply packages/native-utils/tests/sign.test.ts]: done

File: packages/native-utils/tests/sign.test.ts [hunk 6]
```diff
@@ -88,19 +89,86 @@ describe('Sign state', () => {
...
+    }
+
+    const peerState = {
+      ...DEFAULT_STATE,
+      outcome,
+    }
+
+    // First 2 updates
+    currentState.turnNum = 1;
+    peerState.turnNum = 2;
```
[the remark this answers is not in the document]
Comment (isoosiss7) [reply packages/native-utils/tests/sign.test.ts]: My code intended to mimic https://github.com/statechannels/statechannels/blob/85f923c7d51b2ca362537642d6455b35cdfa30c8/packages/nitro-protocol/contracts/ForceMove.sol#L566 but had a bug, fixed

**Comment by isoosiss7** [comment]
> I don't think you're looking for an approval, as you mentioned that this is a draft. For instance, the logic is currently missing the transition validation by the app definition -- I guess that's coming up soon?

For transition validation by app definition, isn't that handled by customer logic in the caller - the library just returns "NeedToCheckApp" to tell the caller to invoke this customer logic?

I will slack you to discuss this next week.

**Review by andrewgordstewart — APPROVED** [review]
Nice work!

**Pull request merged**